src/ppom-controller.ts

```diff
@@ -0,0 +1,387 @@
+ * @property provider - The provider used to create the PPOM instance
+ */
+export class PPOMController extends BaseController<
+  PPOMControllerConfig,
+  PPOMState
+> {
+  /**
+   * Name of this controller used during composition
+   */
+  override name = 'PPOMController';
+
+  #ppom: PPOM | undefined;
+
+  #provider: any;
+
+  #storage: PPOMStorage;
+
+  /*
+   * This mutex is used to prevent concurrent usage of the PPOM instance
+   * and protect the PPOM instance from being used while it is being initialized/updated
+   */
+  #ppomMutex: Mutex;
+
+  /**
+   * Creates a PPOMController instance.
+   *
+   * @param options - Constructor options.
+   * @param options.storageBackend - The storage backend to use for storing PPOM data.
+   * @param options.provider - The provider used to create the PPOM instance.
+   * @param options.chainId - Id of current chain.
+   * @param options.onNetworkChange - Callback tobe invoked when network changes.
+   * @param options.config - The controller configuration.
+   * @param options.state - The controller state.
+   * @returns The PPOMController instance.
+   */
+  constructor({
+    storageBackend,
+    provider,
+    chainId,
+    onNetworkChange,
+    config,
+    state,
+  }: {
+    storageBackend: StorageBackend;
+    provider: any;
+    chainId: string;
+    onNetworkChange: (callback: (chainId: string) => void) => void;
+    config?: PPOMControllerConfig;
+    state?: PPOMState;
+  }) {
+    const defaultConfig = {
+      refreshInterval: DAY,
+    };
+    const defaultState = {
+      lastFetched: 0,
+      versionInfo: [],
+      storageMetadata: [],
+      lastChainId: '',
+      newChainId: chainId,
+    };
+    super(config ?? defaultConfig, state ?? defaultState);
+
+    this.defaultConfig = defaultConfig;
+    this.defaultState = defaultState;
+
+    this.#provider = provider;
+    this.#storage = new PPOMStorage({
+      storageBackend,
+      readMetadata: () => {
+        return this.state.storageMetadata;
+      },
+      writeMetadata: (metadata) => {
+        this.update({ storageMetadata: metadata });
+      },
+    });
+    this.#ppomMutex = new Mutex();
+
+    onNetworkChange((id: string) => {
+      this.update({ newChainId: id });
+    });
+
+    this.initialize();
+  }
+
+  /**
+   * Set the interval at which the ppom version info will be fetched.
+   * Fetching will only occur on the next call to test/bypass.
+   * For immediate update to the ppom lists, call updatePPOM directly.
+   *
+   * @param interval - The new interval in ms.
+   */
+  setRefreshInterval(interval: number) {
+    this.configure({ refreshInterval: interval }, false, false);
+  }
+
+  /**
+   * Determine if an update to the ppom configuration is needed.
+   * The function will return true if
+   * - the chainId has changed
+   * - the ppom is out of date
+   * - the ppom is not initialized.
+   *
+   * @returns True if PPOM data requires update.
+   */
+  async #shouldUpdate(): Promise<boolean> {
```

Comment:
Is it best practice to group all private methods to below all public methods? If only to make it easier to navigate all the public API.